### PR TITLE
docs: consolidate on a simple HACKING.md

### PR DESCRIPTION
**Pull request by sergiusens**
Make it consistent to a single environment and easier to setup.

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: HACKING.md [hunk 1]
```diff
@@ -1,95 +1,69 @@
...
+        libffi-dev \
+        libsodium-dev \
+        libxml2-dev \
+        libxslt-dev
+        libyaml-dev \
+        make \
+        p7zip-full \
+        patchelf \
+        python3-dev \
+        python3-venv \
```
Comment (kyrofa) [inline HACKING.md]: You never actually set up a venv. Either you should do that, or you can get rid of this.

File: HACKING.md [hunk 2]
```diff
@@ -1,95 +1,69 @@
...
+        rpm2cpio \
+        squashfs-tools \
+    lxc exec snapcraft-dev -- sudo -u ubuntu pip3 install -U pip
+    lxc exec snapcraft-dev -- sudo -u ubuntu pip install \
```
Comment (cjp256) [inline HACKING.md]: Default behavior is current working directory is /root. Use -i for sudo. Also use `--user` for pip if installing as user.
```
lxc exec snapcraft-dev -- sudo -i -u ubuntu pip install -U \
    -r snapcraft/requirements.txt \
    -r snapcraft/requirements-devel.txt
```

File: HACKING.md [hunk 3]
```diff
@@ -1,95 +1,69 @@
...
+        -r snapcraft/requirements.txt \
+        -r snapcraft/requirements-devel.txt
+
+Optionally, to quickly try out snapcraft from within the environment:
+
+    lxc exec snapcraft-dev -- sudo -u ubuntu pip install --editable snapcraft
```
Comment (cjp256) [inline HACKING.md]: Trying to install as user, I added `--user`, but I get the following error:
```
Installing collected packages: snapcraft
  Found existing installation: snapcraft 3.5
    Can't uninstall 'snapcraft'. No files were found to uninstall.
  Running setup.py develop for snapcraft
    ERROR: Complete output from command /usr/bin/python3 -c 'import setuptools, tokenize;__file__='"'"'/home/ubuntu/snapcraft/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix=:
    ERROR: usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: -c --help [cmd1 cmd2 ...]
       or: -c --help-commands
       or: -c cmd --help
    
    error: option --user not recognized
    ----------------------------------------
  ERROR: Can't roll back snapcraft; was not uninstalled
ERROR: Command "/usr/bin/python3 -c 'import setuptools, tokenize;__file__='"'"'/home/ubuntu/snapcraft/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps --user --prefix=" failed with error code 1 in /home/ubuntu/snapcraft/
```

File: HACKING.md [hunk 4]
```diff
@@ -1,95 +1,69 @@
...
+    lxc exec snapcraft-dev -- apt update
+    lxc exec snapcraft-dev -- apt install \
+        execstack \
+        g++ \
+        gcc \
+        libapt-pkg-dev \
+        libffi-dev \
+        libsodium-dev \
+        libxml2-dev \
+        libxslt-dev
```
Comment (cjp256) [inline HACKING.md]: Missing trailing -'\'

File: HACKING.md [hunk 5]
```diff
@@ -1,95 +1,69 @@
...
+        libyaml-dev \
+        make \
+        p7zip-full \
+        patchelf \
+        python3-dev \
+        python3-venv \
+        rpm2cpio \
+        squashfs-tools \
+    lxc exec snapcraft-dev -- sudo -u ubuntu pip3 install -U pip
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply HACKING.md]: yes, good catch, the `-i` is something I always use and somehow missed it here

File: HACKING.md [hunk 6]
```diff
@@ -1,95 +1,69 @@
...
+        libxml2-dev \
+        libxslt-dev
+        libyaml-dev \
+        make \
+        p7zip-full \
+        patchelf \
+        python3-dev \
+        python3-venv \
+        rpm2cpio \
+        squashfs-tools \
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply HACKING.md]: good catch, I did a "sort" and didn't check the result.

File: HACKING.md [hunk 7]
```diff
@@ -1,95 +1,74 @@
...
-Install snapcraft (and its dependencies):
+### Testing
 
-    (snapcraft) $ pip install -r requirements.txt .
+See the [Testing guide](TESTING.md).
 
+### Enabling debug output
 
-## Running
+Given that the `--debug` option in snapcraft is reserved for project specific debugging, enabling for the `logger.debug` calls is achieved by setting the "SNAPCRAFT_ENABLE_DEVELOPER_DEBUG" environment variable to a truthful value. Snapcraft's internal tools, e.g.; `snapraftctl` should pick up this environment variable as well.
```
Comment (cjp256) [inline HACKING.md]: s/snapraftctl/snapcraftctl

File: HACKING.md [hunk 8]
```diff
@@ -1,95 +1,74 @@
...
+    lxc exec snapcraft-dev -- sudo -iu ubuntu pip3 install --upgrade --user pip
+    lxc config device add snapcraft-dev snapcraft-project disk source="$PWD" path=/home/ubuntu/snapcraft
+    lxc exec snapcraft-dev -- sudo -iu ubuntu pip install --user \
+        -r snapcraft/requirements.txt \
+        -r snapcraft/requirements-devel.txt
+
+Optionally, to quickly try out snapcraft from within the environment:
+
+    lxc exec snapcraft-dev -- sudo -iu ubuntu bash -c \
+        "cd snapcraft && pip install --no-use-pep517 --user --editable ."
```
Comment (cjp256) [inline HACKING.md]: perhaps just add the (obvious?):
```
lxc exec snapcraft-dev -- sudo -iu ubuntu bash
```

Or append it to the pip install? :) e.g.
```
lxc exec snapcraft-dev -- sudo -iu ubuntu bash -c \
        "cd snapcraft && pip install --no-use-pep517 --user --editable . && bash"
```

**Comment by cjp256** [comment]
Tested out the latest update, works well - thanks!

File: HACKING.md [hunk 9]
```diff
@@ -1,95 +1,69 @@
...
+
+We want to make sure everyone develops using a consistent base, to ensure that these instructions rely on LXD (use whatever is convenient as long as you do not stray away from an Ubuntu 16.04 LTS base)
+
+First we setup LXD:
+
+    sudo snap install lxd
+    sudo lxd init  # If unsure, pick `dir` as the storage backend.
+    sudo adduser "$USER" lxd
+    newgrp lxd
+    lxc create ubuntu:16.04 snapcraft-dev
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply HACKING.md]: init actually

File: tools/environment-setup.sh [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+# Install python dependencies
+lxc exec snapcraft-dev -- sudo -iu ubuntu pip install \
+    -r snapcraft/requirements.txt \
+    -r snapcraft/requirements-devel.txt
+
+# Install the project for quick tests
+lxc exec snapcraft-dev -- sudo -iu ubuntu pip install --editable snapcraft
+
+echo "Environment ready, enter it by running: "
+echo "lxc exec snapcraft-dev -- sudo -iu ubuntu bash"
```
Comment (cjp256) [inline tools/environment-setup.sh]: nit: eof newline?

**Comment by cjp256** [comment]
tested latest update, LGTM

**Pull request merged**